COOKBOOK/EXTERNAL_COMPLETERS.MD Add filepath escape troubleshooting (#1844)
nushell/nushell.github.io: cookbook/external_completers.md

    fish --command $'complete "--do-complete=($spans | str join " ")"'
    | from tsv --flexible --noheaders --no-infer
    | rename value description
    | update value {

>         if ($in | path exists) {$'"($in | str replace "\"" "\\\"" )"'} else {$in}

Reviewer: Not a huge deal but using single quotes helps avoid the escapes
```suggestion
        if ($in | path exists) { $'"($in | str replace '"' '\"' )"' } else { $in }
```

Author: it uses single quotes for the substitute `$'` so that won't work

Reviewer: Whoops, didn't realize that

Author: yeah. it would be really nice if 
```
$` () `
```
was a thing

- The fish completer will return lines of text, each one holding the `value` and `description` separated by a tab. The `description` can be missing, and in that case there won't be a tab after the `value`. If that happens, `from tsv` will fail, so we add the `--flexible` flag.
- The output of the fish completer does not contain a header (name of the columns), so we add `--noheaders` to prevent `from tsv` from treating the first row as headers and later give the columns their names using `rename`.
- `--no-infer` is optional. `from tsv` will infer the data type of the result, so a numeric value like some git hashes will be inferred as a number. `--no-infer` will keep everything as a string. It doesn't make a difference in practice but it will print a more consistent output if the completer is ran on it's own.

> - Since fish only supports POSIX style escapes for file paths (`file\ name.txt`, etc.), file paths completed by fish will not be quoted or escaped properly on external commands. Nushell does not have a way to natively parse POSIX escapes, so we need to do this conversion manually such as by testing if the items are valid paths as shown in the example. This simple approach is imperfect, but it should cover the 99.9%.

Reviewer: Minor wording change (as far as I know, we don't plan to handle `\ `, so it's more that Nushell doesn't do it than that it hasn't been implemented yet)
```suggestion
- Since fish only supports POSIX style escapes for file paths (`file\ name.txt`, etc.), file paths completed by fish will not be quoted or escaped properly on external commands. Nushell does not parse POSIX escapes, so we need to do this conversion manually such as by testing if the items are valid paths as shown in the example. This simple approach is imperfect, but it should cover 99.9% of use cases.
```

Author: lgtm

}
```

> ### File path escapes

Reviewer: Imperfect though this may be, I think it would be good to include your unescaping code right in the `$fish_completer` definition above, rather than in the troubleshooting section, especially if this is a common issue Fish completer users run into.

Author: It really applies to all completers that don't directly support nushell syntax, not just fish. Fish is just the common victim because its the most popular

Reviewer: I see. I'd still rather have the Fish completer above include the unescaping logic (also the Zoxide completer), so people don't have to come back here to figure out why their completer isn't escaping stuff correctly.

And in addition to that, you can leave this bit in the troubleshooting section in addition to that, in case people use another completer. I'd rename `fish_completer_escaped` to `completer_escaped` and so on in that case, though.

Author: ah if zoxide is good and its just fish sure
